### DISCUSSION TITLE
export pdf and csv

### DISCUSSION THREAD

**Discussion by ashkankiani** [TanStack/table#4402] [Q&A]
It is a good library, but it needs to add data output to the system.
How to get output from data. Settings for raw or modified data output.
For example, a data is shown with true, while we mean "paid". And we want the output to be shown like this.

**Comment by KevinVandy** [comment]
You can easily hook up all of the getRowModel methods to a data export library. You can kind of get an idea for how to do this in an example of this library that is built on top of TanStack Table. Pretty much just connect whichever row model you want to a custom export library like `export-to-csv`

https://www.material-react-table.com/docs/examples/data-export

**Reply by Shaglock** [reply]
Hello @KevinVandy, thank you for the answer and the example.

I was wondering if is there any way to reuse the cells from columns definition, for example I have a formatted values in the cells, and I want to export the rows with those formatted values? 

For example, I have this column. 
```ts
  columnHelper.accessor('someArrayField', {
    cell: (info) =>
      info
        .renderValue()
        ?.map((v) => v.name)
        .join(', ') ?? null,
  }),
  ```
  Is there any way to reuse this `cell` definition for export? I guess doing something like `row.getAllCells().map(c => c.column.columnDef.cell(c.getContext()))` would work but feels really off.
  
  Maybe I can also remove this cell and change accessor to accessor function like this
  ```ts
  columnHelper.accessor((row)=>row.someArrayField?.map(v=>v.name).join(', ') ?? '', {
  id: 'someArrayField'
  }),
  ```
  
  and then use renderValue instead but still feels like a lot of overhead
  ```ts
  row.getVisibleCells().map((cell) => cell.renderValue())
   ```
 Or maybe the second approach is the best choice we have?

**Reply by evbo** [reply]
@KevinVandy I'm a bit confused, it seems you're suggesting "do not use react table V8 for data export"?
When I read the example you linked, it exports data completely external from react table. They simply store a reference to all of the data.

For instance, here is their handler, with it's own reference to the data (totally separate from the table - how could this support pagination, etc.?):
```javascript
const handleExportData = () => {
    csvExporter.generateCsv(data);
  };
```

I was expecting a similar functionality V7 provided. Does that exist? Is that what you were referring to with `getRowModel`? Can you clarify more on that approach?

**Reply by KevinVandy** [reply]
Just 1 of 4 examples used the data. The other 3 looks at the rowModels

**Reply by evbo** [reply]
@KevinVandy thanks, I couldn't find any other examples on that page. Where did you see the 3 others?

**Reply by KevinVandy** [reply]
```jsx
const handleExportRows = (rows: MRT_Row<Person>[]) => {
    csvExporter.generateCsv(rows.map((row) => row.original));
  };

  const handleExportData = () => {
    csvExporter.generateCsv(data);
  };
```

The first function in that example that is used by the other 3 buttons gets data from the row model. It is all there

**Reply by KevinVandy** [reply]
So for example, if you only wanted to export selected rows from the table, you could have a function like this

```ts
const handleExportSelectedRows = () => {
    csvExporter.generateCsv(table.getSelectedRowModel().rows.map((row) => row.original));
};
```

**Comment by naveen1337** [comment]
when a client side pagination implemented through react table, I solved like that,

```ts
export function loopReactTablePaginationData(tableInstance: any) {
  try {
    const stringify = JSON.stringify(tableInstance);
    const clone = JSON.parse(stringify);

    const rows: any = [];
    for (let i = 0; i < clone.pageCount; i++) {
      clone.page.map((item: any) => {
        rows.push(item.original);
      });
      if (clone.canNextPage) {
        clone.nextPage();
      }
    }
    return rows;
  } catch (err) {
    console.log(err);
    return [];
  }
}
```

but the recommended to way to export data just do it on server side

**Comment by ashkankiani** [comment]
Thank you very much. Thank you for taking the time to guide me. Your
advice was very helpful for me.

On Mon, Sep 26, 2022 at 3:20 PM Naveen 1337 ***@***.***>
wrote:

> when a client side pagination implemented through react table, I solved
> like that,
>
> export function loopReactTablePaginationData(tableInstance: any) {
>   try {
>     const stringify = JSON.stringify(tableInstance);
>     const clone = JSON.parse(stringify);
>
>     const rows: any = [];
>     for (let i = 0; i < clone.pageCount; i++) {
>       clone.page.map((item: any) => {
>         console.log(item);
>         rows.push(item.original);
>       });
>       if (clone.canNextPage) {
>         clone.nextPage();
>       }
>     }
>     return rows;
>   } catch (err) {
>     console.log(err);
>     return [];
>   }}
>
> but the recommended to way to export data just do it on server side
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/TanStack/table/discussions/4402#discussioncomment-3733261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AF4A6SO3AN3WQDN34KYODZ3WAGEYBANCNFSM6AAAAAAQUQDG2M>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>


-- 
---------------------------------------------------------------
Best Regards.
Ashkan Kiani
Web Developer and Web Designer
